### PR TITLE
Secure backup

### LINKED ISSUE #1574: [Secure Backup] Chat backup settings section
From element-x-ios created by [stefanceriu](https://github.com/stefanceriu): vector-im/element-x-ios#1899  

* A new Chat backup entry should be visibile in the application settings
    * If this particular session isn't verified then Chat backup options aren't available
* Backup and recovery should correctly display their current state
* There should be a link to a page that describes `chat backup` in more details

**If** recovery isn't setup then badges should be shown throughout the app (user avatar, settings option, secure backup entry) to guide the user towards setting up a recovery

### Designs
https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12124-116601&mode=design&t=7YPtBpnsTXrsNrdP-0

### LINKED ISSUE #1575: [Secure backup] Last session logout without recovery
From element-x-ios created by [stefanceriu](https://github.com/stefanceriu): vector-im/element-x-ios#1903  

When attempting to sign out from their last session the user should be prompted to setup recovery so they don't lose access to their encrypted messages

### Design

https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12495-22145&mode=design&t=lxXaFMJStZVxqef3-0

### LINKED ISSUE #1576: [Secure backup] Timeline indicators
From element-x-ios created by [stefanceriu](https://github.com/stefanceriu): vector-im/element-x-ios#1902  

With key backup available there might now be multiple reasons why messages cannot be decrypted (yet)

* A timeline banner should be shown if the session is not verified
* Another banner should be shown if the session is verified but key backup is disabled
* Another banner should be shown if the session is verified, key backup is available but the keys haven't been downloaded yet

### Designs
https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12283-143652&mode=design&t=lxXaFMJStZVxqef3-0

### LINKED ISSUE #1577: [Secure backup] Recovery screen
From element-x-ios created by [stefanceriu](https://github.com/stefanceriu): vector-im/element-x-ios#1901  

Enabling, changing or fixing recovery should be possible from a new screen accessible through the, also new, chat backup settings section.

* Once a new recovery key has been generated the application shouldn't allow the user to continue until the key has either been copied to the clipboard or saved. A confirmation prompt should also be shown. We need to do our best to ensure the user doesn't lose it.
* We should use the system share sheet for "saving" the key but we should at least attempt to remove less safe options from there. This is not something you want shared on Facebook.

### Designs
https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12283-150028&mode=design&t=lxXaFMJStZVxqef3-0

https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12283-150036&mode=design&t=lxXaFMJStZVxqef3-0

https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12495-22844&mode=design&t=lxXaFMJStZVxqef3-0

### LINKED ISSUE #1578: [Secure Backup] Backup screen
From element-x-ios created by [stefanceriu](https://github.com/stefanceriu): vector-im/element-x-ios#1900  

Enabling or disabling key backup should be possible from a new screen accessible through the, also new, chat backup settings section.

### Designs
https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12124-116601&mode=design&t=7YPtBpnsTXrsNrdP-0

https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=12998-37515&mode=design&t=lxXaFMJStZVxqef3-0

### PR DESCRIPTION
**Pull request by bmarty**
*The original big PR has been split to smaller one. This PR now include bug fixes and improvement.*

 
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

First implementation of secure backup:
- add a way to create the recovery key
- add a way to change the recovery key
- add a way to disable the keybackup
- add a way to enable the keybackup
- update the logout flow

Figma: https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?node-id=12998%3A32971&mode=dev

## Motivation and context

Fixes #1574
Fixes #1575
Fixes #1577
Fixes #1578

## Screenshots / GIFs

See screenshot in the PR

Demo (including bugs)

SecureBackupDemo

## Tests

- Login to a session and verify it then go to the new `Chat backup` setting entry and test all the new screens.

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s):

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-x-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR

### CODE REVIEW

File: features/lockscreen/impl/src/main/kotlin/io/element/android/features/lockscreen/impl/unlock/PinUnlockView.kt [hunk 1]
```diff
@@ -229,7 +229,8 @@ private fun PinUnlockHeader(
         val subtitle = if (state.showWrongPinTitle) {
             pluralStringResource(id = R.plurals.screen_app_lock_subtitle_wrong_pin, count = state.remainingAttempts, state.remainingAttempts)
         } else {
-            stringResource(id = R.string.screen_app_lock_subtitle)
+            TODO()
+            // stringResource(id = R.string.screen_app_lock_subtitle)
```
Comment (bmarty) [inline features/lockscreen/impl/src/main/kotlin/io/element/android/features/lockscreen/impl/unlock/PinUnlockView.kt:233]: Need other PR with PIN...

**Review by jmartinesp — APPROVED** [review]
LGTM! 

Also, not directly related to this PR, but is it normal that I still see the encrypted history banner in the timeline after verifying my account and even confirming the recovery key?

**Comment by bmarty** [comment]
(rebased and screenshooting again)

**Comment by bmarty** [comment]
> Also, not directly related to this PR, but is it normal that I still see the encrypted history banner in the timeline after verifying my account and even confirming the recovery key?

Not really, I will investigate this.

**Pull request merged**